File: Scala Kompics/src/se/sics/kompics/scala/performancetest/kompics/SimpleBroadcast.scala
package se.sics.kompics.scala.performancetest.kompics

import se.sics.kompics.scala.ComponentDefinition
import se.sics.kompics.network.Network
import se.sics.kompics.address.Address
import se.sics.kompics.performancetest.launcher.ScalaTopology
import se.sics.kompics.performancetest.launcher.ScalaTopology._
import se.sics.kompics.performancetest.scenario.TopologyView
import se.sics.kompics.scala.Init
import se.sics.kompics.network.Message
import se.sics.kompics.network.Transport


case class SBMessage(src: Address, dest: Address, msg: Any) extends Message(src, dest, Transport.TCP)

class SimpleBroadcast extends ComponentDefinition {
	val beb = ++(BasicBroadcast);
	val network = --(classOf[Network]);

	private var topology: TopologyView = null;

	def self: Address = topology.getSelfAddress();
	def neighbours: Set[Address] = topology.neighbours;

	ctrl uponEvent {
		case Init(topology: TopologyView) => { () =>
			this.topology = topology;
		}
	}

	beb uponEvent {
		case SBSend(msg) => { () =>
			neighbours foreach {
				q => trigger(SBMessage(self, q, msg), network);
			}
		}
	}

	network uponEvent {
		case SBMessage(src, dest, msg) => { () =>
			trigger(SBDeliver(src, msg), beb);
		}
	}
}

File: Java Kompics/src/se/kth/ii2202/experiments/beb/SimpleBroadcast.java
package se.kth.ii2202.experiments.beb;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Handler;
import se.sics.kompics.Negative;
import se.sics.kompics.Positive;
import se.sics.kompics.address.Address;
import se.sics.kompics.network.Network;
import se.sics.kompics.performancetest.scenario.TopologyView;
    
/**
 *
 * @author carbone
 */
public class SimpleBroadcast extends ComponentDefinition {

    private static final Logger logger = LoggerFactory.getLogger(SimpleBroadcast.class);
    Negative<BestEffortBroadcast> beb = provides(BestEffortBroadcast.class);
    Positive<Network> network = requires(Network.class);
    Address selfAddr;
    TopologyView topology;

    public SimpleBroadcast() {
        subscribe(onInit, control);
        subscribe(onUnBroadcast, beb);
        subscribe(onMessageReceived, network);
    }
    Handler<BEBInit> onInit = new Handler<BEBInit>() {
        @Override
        public void handle(BEBInit e) {
            SimpleBroadcast.this.topology = e.getTopology();
            SimpleBroadcast.this.selfAddr = topology.getSelfAddress();
        }
    };
    Handler<BroadcastRequest> onUnBroadcast = new Handler<BroadcastRequest>() {
        @Override
        public void handle(BroadcastRequest e) {
        	//System.out.println("Broadcasting message of type " + e.getMsg().getClass().getCanonicalName());
            for (Address addr : topology.getNeighbours()) {
                trigger(new BroadcastMessage(selfAddr, addr, e.getMsg()), network);
            }
        }
    };
    Handler<BroadcastMessage> onMessageReceived = new Handler<BroadcastMessage>() {
        @Override
        public void handle(BroadcastMessage e) {
            trigger(e.getMsg(), beb);
        }
    };
}
